sympy/solvers/ode/systems.py

```diff
@@ -1099,40 +1110,64 @@ def _ode_component_solver(eqs, funcs, t, const_idx=0):
     match['const_idx'] = const_idx
     match['t'] = t
 
-    if match['is_linear']:
+    if match.get('is_linear', False):
         return [_linear_ode_solver(match)], const_idx + len(eqs)
     if match.get('is_implicit', False):
         canon_eqs = match['canon_eqs']
         sols = []
         for canon_eq in canon_eqs:
-            sol = _ode_component_solver(canon_eq, funcs, t, const_idx=const_idx)
+            sol, const_idx = _ode_component_solver(canon_eq, funcs, t, const_idx=const_idx)
             if sol is not None:
-                sols.append(sol)
-            const_idx += len(canon_eq)
+                sols += sol
         return sols, const_idx
 
     # To add non-linear case here in future
 
     return None
 
 
-def dsolve_system(eqs, funcs, t):
+def dsolve_system(eqs, funcs=None, t=None):
     r"""
 
     """
 
     # Note: To add proper preprocessing function
     # where the function can get the equations from
     # the expressions.
+
+    eqs = _preprocess_eqs(eqs)
+
     if any(not isinstance(eq, Eq) for eq in eqs):
         raise ValueError(filldedent('''
             List of equations should be passed. The input is not valid.
         '''))
 
-    # Note: To add proper error for passing incorrect funcs
-    # input later.
+    # Note: This error checking has to be improved later.
+    if funcs is not None and not isinstance(funcs, list):
+        raise ValueError(filldedent('''
+            Input to the funcs should be a list of functions.
+        '''))
 
-    sol = []
+    # Note: This is added to solve a major problem encountered.
+    # Might be best to make a function for this function
+    # extraction later.
```

Comment:
What is the major problem?

Reply:
Well, I wanted to add `dsolve_system` call in `dsolve` itself. But if I did that, then there is a high chance that a user will not pass the `funcs` which would make us depend on `classify_sysode`. Now, `classify_sysode` calls our matching function and returns its output. Since our `dsolve_system` is made so that even an end-user can call it, we are not passing the match duct. This would mean we would have to call our matching function twice. 

The above was the problem. To solve that, I simply added this `funcs` extractor code. This removed the risk of calling the matching function twice.

Reply:
It's probably worth making this code for guessing the functions into a separate function that can be used in both places.